File: client/src/app/noticias/services/periodico-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import {
  Periodico,
  CrearPeriodicoDTO,
  ActualizarPeriodicoDTO,
} from '@/noticias/interfaces/Periodico';

@Injectable({
  providedIn: 'root',
})
export class PeriodicoService {
  private http = inject(HttpClient);

  private readonly apiUrl = '/api/periodicos';

  public listar(): Observable<Periodico[]> {
    return this.http.get<Periodico[]>(this.apiUrl);
  }

  public obtener(id: number): Observable<Periodico> {
    return this.http.get<Periodico>(`${this.apiUrl}/${id}`);
  }

  public crear(dto: CrearPeriodicoDTO): Observable<Periodico> {
    return this.http.post<Periodico>(this.apiUrl, dto);
  }

  public eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  public actualizar(id: number, dto: ActualizarPeriodicoDTO): Observable<Periodico> {
    return this.http.put<Periodico>(`${this.apiUrl}/${id}`, dto);
  }
}

File: api/src/main/java/info/noticiasweb/periodicos/services/PeriodicoService.java
package info.noticiasweb.periodicos.services;

import java.util.ArrayList;
import java.util.List;

import info.noticiasweb.periodicos.dao.PeriodicoDAO;
import info.noticiasweb.periodicos.dtos.ActualizarPeriodicoDTO;
import info.noticiasweb.periodicos.dtos.CrearPeriodicoDTO;
import info.noticiasweb.periodicos.dtos.PeriodicoDTO;
import info.noticiasweb.periodicos.models.Periodico;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

@ApplicationScoped
public class PeriodicoService {
    @Inject
    PeriodicoDAO periodicoDAO;

    public List<PeriodicoDTO> listar() {
        List<PeriodicoDTO> periodicoDTOs = new ArrayList<>();
        for (Periodico p : this.periodicoDAO.listar()) {
            periodicoDTOs.add(new PeriodicoDTO(p));
        }
        return periodicoDTOs;
    }

    public PeriodicoDTO crear(CrearPeriodicoDTO dto) {
        return new PeriodicoDTO(this.periodicoDAO.crear(dto));
    }

    public void eliminar(Long id) {
        this.periodicoDAO.eliminar(id);
    }

    public PeriodicoDTO obtener(Long id) {
        return new PeriodicoDTO(this.periodicoDAO.obtener(id));
    }

    public PeriodicoDTO actualizar(Long id, ActualizarPeriodicoDTO dto) {
        return new PeriodicoDTO(this.periodicoDAO.actualizar(id, dto));
    }

}
